### PR TITLE
Bug 1709575: account for the jenkins SA certificate differing from th…

### PR DESCRIPTION
**Pull request by gabemontero**
…e oauth server's router certificate (where SSL handshake errors would then occur with jenkins SA certificate) by falling back to the JVM's default keystore

@openshift/sig-developer-experience @jupierce @enj fyi

### CODE REVIEW

File: src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java [hunk 1]
```diff
@@ -168,6 +168,13 @@
      * Will get repopulated on restart
      */
     private static HttpTransport transport;
+    
+    /**
+     * transport that will only leverage the JVMs default keystore (allow for the jenkins SA cert
+     * and the oauth server router cert varying such that SSL handshakes will fail
```
Comment (adambkaplan) [inline src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java]: nit - missing close paren

File: src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java [hunk 2]
```diff
@@ -574,19 +613,41 @@ public void initialize(HttpRequest request)
...
+            if (rc != 200)
+                LOGGER.info("OpenShift OAuth the attempt to get the server version request got an unexpected return code: " + rc);
+            OpenShiftVersionInfo version = response.parseAs(OpenShiftVersionInfo.class);
+            if (version != null && version.major != null && version.major.equals("1")) {
+                if (version.minor.length() > 2) {
+                    String minor = version.minor.substring(0, 2); // per javadoc end index is not inclusive
+                    int m = Integer.parseInt(minor);
+                    if (m <= 11) {
+                        // 3.x cluster
+                        LOGGER.info("OpenShift OAuth the server is 3.x, specifically " + version.toString());
```
Comment (adambkaplan) [inline src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java]: Are the version numbers for the OpenShift OAuth server completely independent from the product version?

File: src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftVersionInfo.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.openshift.jenkins.plugins.openshiftlogin;
```
Comment (adambkaplan) [inline src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftVersionInfo.java]: Side note - can't we have a top-level LICENSE file? (if so I'd put this in a separate PR so someone in legal can weigh in)

File: src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftVersionInfo.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2004-2009, Sun Microsystems, Inc., Kohsuke Kawaguchi
```
Comment (adambkaplan) [inline src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftVersionInfo.java]: I'd strike this, as Sun has nothing to do with this file.

File: src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java (hunk 2 above)
Comment (jupierce) [reply src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java]: What is being checked here is the underlying kubernetes version. It is effectively independent with 4.0.

File: src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java (hunk 1 above)
Comment (gabemontero) [reply src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java]: part of next push

File: src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftVersionInfo.java (hunk 4 above)
Comment (gabemontero) [reply src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftVersionInfo.java]: there are refs to classes constructed at that time  ... I remember having this discussion with Clayton wrt when this plugin first came into being

Not going to diverge at this point

File: src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftVersionInfo.java (hunk 3 above)
Comment (gabemontero) [reply src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftVersionInfo.java]: devtools can pursue that at this point :-) .... "meets min" wrt effort @adambkaplan :-)

File: src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java (hunk 2 above)
Comment (gabemontero) [reply src/main/java/org/openshift/jenkins/plugins/openshiftlogin/OpenShiftOAuth2SecurityRealm.java]: Yep it is just a debug marker to note which path wrt transports, oauth endpoints, and certs is being taken

OAuth behavior is different between 3.x and 4.x @adambkaplan , so we vary accordingly.  Goal is to be able to use 1 version of the login plugin across 3.x and 4.x to simplify support.

**Comment by gabemontero** [comment]
comment correction nit pushed @adambkaplan ...explanations/response to the other comments in line with various conversations started

ptal

**Comment by gabemontero** [comment]
I'll be cutting v1.0.17 of the login plugin for this fix
